Fix data directory deleted on migrations

lucernae opened the pull request:
Fix for https://github.com/kartoza/docker-postgis/issues/226

Making sure conscious decision of reinitializing data directory if the data already exists.

Propose a two stage decisions:

1. Setting EXISTING_DATA_DIRECTORY to true (default value) will use existing data directory. The default value is trying to make sure on unknowing upgrade process, the data directory should not be considered empty.
2. Delete/Add `.postgresql.init.lock` to indicate the intent to reinitialize database. However in previous image, this lock file doesn't exists. This is why we need two stage decisions, so the DBAdmin can confirm the changes.

If either of this decisions were mismatched, service will hang. So you can go inside the shell and confirm the decision by creating/deleting the init lock file and data directory.

lucernae committed:
Fix data directory deleted on migrations

lucernae commented:
Also fix for https://github.com/kartoza/docker-postgis/issues/227

NyakudyaA commented:
> Fix for #226
> 
> Making sure conscious decision of reinitializing data directory if the data already exists.
> 
> Propose a two stage decisions:
> 
>     1. Setting EXISTING_DATA_DIRECTORY to true (default value) will use existing data directory. The default value is trying to make sure on unknowing upgrade process, the data directory should not be considered empty.
> 
>     2. Delete/Add `.postgresql.init.lock` to indicate the intent to reinitialize database. However in previous image, this lock file doesn't exists. This is why we need two stage decisions, so the DBAdmin can confirm the changes.
> 
> 
> If either of this decisions were mismatched, service will hang. So you can go inside the shell and confirm the decision by creating/deleting the init lock file and data directory.
@lucernae I have had a look at this. I think we might be over-engineering the solution here.

* I can't just spin up a DB instance without having to do some extra configurations for now.
![data](https://user-images.githubusercontent.com/2510900/76507524-9f8c6f00-6455-11ea-9794-c8e900232708.png)

NyakudyaA commented on setup-database.sh:
```diff
@@ -8,22 +8,41 @@ SETUP_LOCKFILE="${DATADIR}/.postgresql.init.lock"
-if [[ -z "${EXISTING_DATA_DIR}" ]]; then \
-    if [[ ! -f "${SETUP_LOCKFILE}" ]]; then
+if [[ "${EXISTING_DATA_DIR}" == 'FALSE' ]]; then
+    # Only attempt reinitializations if ${DATADIR} doesn't exists
```
When PostgreSQL is installed it initializes the data directory at /var/lib/postgresql/12/main. Previously we used to check if the data directory exists and hence the https://github.com/kartoza/docker-postgis/issues/201 which I attempted to solve. 

I think we can either check to see what files exists when a DB is initialized vs an empty DB.
That would be a simpler solution to handle the upgrade situation but still also let first-time users spin up the image without having to check the logs

lucernae replied:
Just CMIIW, in earlier solution new COLLATION is applied by reinitializing a new DB?
Is there no way to change COLLATION without destroying existing one?
If we can find those, then we don't need init lock file because as you said, the data dir was initialized by PostgreSQL from inside the image, even before entrypoint script is running. We can't intercept those.

NyakudyaA replied:
During initialization PostgreSQL creates a directory ie
![du-sh-default-12](https://user-images.githubusercontent.com/2510900/76541163-d2515a00-648b-11ea-9531-0f823cccb092.png)

I think we should be able to figure out how to compare this data directory to the one where a running instance is so that we can determine if it's the first time running the image or it's an existing image

NyakudyaA replied:
@lucernae When Postgres is installed it initializes the data directory and creates the postgresql.conf. When our scripts run it populates postgresql.conf. We could run the following
```

   # Search for custom settings defined in /setup-conf.sh
    if  grep -Fxq "xmloption = 'document'" ${CONF} ;then \
   echo "PostgreSQL has been initialised before"
else
    # No content yet - first time pg is being run!
    # No Replicate From settings. Assume that this is a master database.
    # Initialise db
    echo "Initializing Postgres Database at ${DATADIR}"
    rm -rf ${DATADIR}/*
    chown -R postgres /var/lib/postgresql
    su - postgres -c "$INITDB -U postgres -E ${DEFAULT_ENCODING} --lc-collate=${DEFAULT_COLLATION} --lc-ctype=${DEFAULT_CTYPE} --wal-segsize=${WAL_SEGSIZE} -D ${DATADIR}"
fi

```
We can even then drop the $SETUP_LOCKFILE and ${EXISTING_DATA_DIR}

lucernae commented:
> @lucernae I have had a look at this. I think we might be over-engineering the solution here.
> I can't just spin up a DB instance without having to do some extra configurations for now.

@NyakudyaA Frankly this is because we uses EXISTING_DATA_DIR and setup lockfile, that's why I implement it like this. I think we just need to choose either one.

To spin up new DB instance, use EXISTING_DATA_DIR true and put a lockfile manually. If this is not convenient, then I don't see any reason at all to have a lockfile.
We could use EXISTING_DATA_DIR default to true, so new instance will run without problem, as always (no need for a lockfile).
Then if you need to reinitialize, you had set EXISTING_DATA_DIR to false and put logic that does (when the flag is false):

1. if the data dir is empty, start reinitializing
2. if the data dir is not empty, warn the user to manually delete the data dir (do not auto delete)
3. Again, we don't have to check for lockfile, because now we have an explicit EXISTING_DATA_DIR flag.

Why we need a lock file in the first place?
Maybe there are some use case that we need to accomodate by introducing a lock file?

NyakudyaA commented:
> > @lucernae I have had a look at this. I think we might be over-engineering the solution here.
> > I can't just spin up a DB instance without having to do some extra configurations for now.
> 
> @NyakudyaA Frankly this is because we uses EXISTING_DATA_DIR and setup lockfile, that's why I implement it like this. I think we just need to choose either one.
> 
> To spin up new DB instance, use EXISTING_DATA_DIR true and put a lockfile manually. If this is not convenient, then I don't see any reason at all to have a lockfile.
> We could use EXISTING_DATA_DIR default to true, so new instance will run without problem, as always (no need for a lockfile).
> Then if you need to reinitialize, you had set EXISTING_DATA_DIR to false and put logic that does (when the flag is false):
> 
>     1. if the data dir is empty, start reinitializing
> 
>     2. if the data dir is not empty, warn the user to manually delete the data dir (do not auto delete)
> 
>     3. Again, we don't have to check for lockfile, because now we have an explicit EXISTING_DATA_DIR flag.
> 
> 
> Why we need a lock file in the first place?
> Maybe there are some use case that we need to accomodate by introducing a lock file?

When a user grabs the image the first time he doesn't have a default data directory so it reinitializes and creates a data dir with the correct collation. A lock file is created to show that the data directory has been reinitalized.

If a user is upgrading a running image we need a simpler check. We need to compare the contents of the data directory that is generated by PostgreSQL on install vs the one for a running container. This will make life easy for everyone.  This way we can even remove the flag existing data directory and have all the logic in the script

lucernae commented:
Hi @NyakudyaA , I'm just able to working on this again today.
Continuing our discussion:

To make it simple, how about this one:

1. Just do a simple check, if the directory does exists, then just uses it
2. If we need to differentiate between `datadir that is created by PostgreSQL` vs `datadir that is maintained by user` we just need to add a metadata file from the docker recipe. So, if user is using datadir from postgres, the file will exists. But this fact should not determine how entrypoint script will delete the datadir or not.
3. Let's just delete environment variable EXISTING_DATA_DIR, rather change it into FORCE_REINITIALIZATION or something (default to False). This is to clearly express the intent, because.
4. If the datadir is not from the same major version of postgres, entrypoint should fail and leave it hanging so the container is up, but unhealthy. That way, user can get inside the container and fix the conflicts themselves (use another dir or change image version).


EDIT:
For no. 3, maybe `RECREATE_DATADIR` is more meaningful.

NyakudyaA commented:
> Hi @NyakudyaA , I'm just able to working on this again today.
> Continuing our discussion:
> 
> To make it simple, how about this one:
> 
>     1. Just do a simple check, if the directory does exists, then just uses it

> 
>     2. If we need to differentiate between `datadir that is created by PostgreSQL` vs `datadir that is maintained by user` we just need to add a metadata file from the docker recipe. So, if user is using datadir from postgres, the file will exists. But this fact should not determine how entrypoint script will delete the datadir or not.

We need to chat about what you think is the data dir created by Postgres vs data directory maintained. Im my mind the datadir created by postgres was used to initialize everything but now due to this setup a user actually initialises the datadir based on locales he/she is using
> 
>     3. Let's just delete environment variable EXISTING_DATA_DIR, rather change it into FORCE_REINITIALIZATION or something (default to False). This is to clearly express the intent, because.

Yes, no problem with changing the variable
> 
>     4. If the datadir is not from the same major version of postgres, entrypoint should fail and leave it hanging so the container is up, but unhealthy. That way, user can get inside the container and fix the conflicts themselves (use another dir or change image version).

I am very much against leaving the container hanging. It should handle both scenarios as before.
> 
> 
> EDIT:
> For no. 3, maybe `RECREATE_DATADIR` is more meaningful.

Just FYI

![fyi](https://user-images.githubusercontent.com/2510900/78121147-4f2c7f80-740b-11ea-96a4-fe449d6d2ae4.png)

Image 12.0 still uses the old logic - it doesn't delete the data.
Image 12.1 uses the new logic where you need to use the variable EXISTID_DATA_DIR
Image 11.0-2.5 still uses the old logic - it doesn't delete the data.
Image 11.5-2.5 uses the new logic where you need to use the variable EXISTID_DATA_DIR

lucernae commented:
@NyakudyaA Just to be clear first, I'm not going to merge this PR as it is.

> We need to chat about what you think is the data dir created by Postgres vs data directory maintained. Im my mind the datadir created by postgres was used to initialize everything but now due to this setup a user actually initialises the datadir based on locales he/she is using

Yes, there's no problem in letting postgres entrypoint script initializing the datadir according to the locale the user chose, we gave the option as `RECREATE_DATADIR` explicitly so we can delete the current datadir and make new. So, can I say this case will be handled?
To be more explicit:

# Datadir precreated by Postgres

In Dockerfile recipe (image creation), by default DATADIR is in `/var/lib/postgresql/12/main` . So, the directory has content (in the image). When user uses named-volume and bind into this default DATADIR: if the named-volume is empty, the existing content is copied into the named volume (by docker automatically). So the default collation/settings is used and can't be changed. The entrypoint script will skip initialization phase because DATADIR already had content.

# Datadir maintained by user

If user reuse existing named-volume that already has DATADIR content, then we should not delete it. The container entry point script can only delete it if RECREATE_DIR is true (the default will be false). DATADIR maintained by user can apply in many scenarios such as:

- User uses our image initially with empty named-volume. At the end, his named-volume has content. If he change image tag but keep the major version (e.g postgis:12.0 into postgis:12.1), he should be able to use the same datadir in his named-volume.

- User migrated his own datadir from somewhere and mount the volume as the container's DATADIR. If it is the same major version, it should work and we don't delete it.

- User migrated his own datadir, or uses initialized named-volume, but from a different postgres/postgis major version. In this case, we should not delete datadir. But rather, hang gracefully with nice log that says why we can't use the datadir (the error message from postgres). It needs to hang so user can get into the container and resolves it. If we don't make it hang, the container will be in an endless loop of restart.

# User gave non default DATADIR locations

If the user set DATADIR in a non default locations (other than `/var/lib/postgresql/12/main`), with empty named-volume, then the DATADIR is empty. So entrypoint must initialize DATADIR according to the chosen conf.

So, to comment on your statement here:

> I am very much against leaving the container hanging. It should handle both scenarios as before.

We're not going to use the current commits in this PR, it will be updated according to the spec that we will agree. What I propose is for the image to handle as above (and then you can add if I miss something). If the scenario is the DATADIR is not empty and postgres can't start, then the container should hang so user can read the logs and resolves the situation. Because we don't know exactly how to resolve it (should we delete the DATADIR or not). User had to give explicit command or settings like RECREATE_DATADIR.

> Let's just delete environment variable EXISTING_DATA_DIR, rather change it into FORCE_REINITIALIZATION or something (default to False). This is to clearly express the intent, because: If the datadir is not from the same major version of postgres, entrypoint should fail and leave it hanging so the container is up, but unhealthy. That way, user can get inside the container and fix the conflicts themselves (use another dir or change image version).

This statement above should have been a single point of idea, my bad, it was separated as point 3 and 4.

To compare it in a table, the difference will be:

| No | Before | After |
| --- | --- | --- |
| 1 | Entry point script delete DATADIR if lock file does not exists | Entry point script will delete DATADIR only if RECREATE_DATADIR is set to true |
| 2 | Entry point script will create lock file when DATADIR is initialized | No lockfile |
| 3 | Entry point script will still reinitialize DATADIR at first run, because lock file doesn't exists even if using DATADIR prebuilt by image | Entry point script will only initialize if DATADIR is empty or RECREATE_DATADIR is true |
| 4 | EXISTING_DATADIR is used as flag to tell that DATADIR exists | Not using EXISTING_DATADIR because it had no purpose in this implementation |
| 5 | EXISTING_DATADIR default to False, so user who upgrade the image pre EXISTING_DATADIR options will not have lock file, and it's data were deleted | Never delete data unless explicit consent as settings were given or user delete it manually |

NyakudyaA commented:
> @NyakudyaA Just to be clear first, I'm not going to merge this PR as it is.
> 
> > We need to chat about what you think is the data dir created by Postgres vs data directory maintained. Im my mind the datadir created by postgres was used to initialize everything but now due to this setup a user actually initialises the datadir based on locales he/she is using
> 
> Yes, there's no problem in letting postgres entrypoint script initializing the datadir according to the locale the user chose, we gave the option as `RECREATE_DATADIR` explicitly so we can delete the current datadir and make new. So, can I say this case will be handled?
> To be more explicit:
> # Datadir precreated by Postgres
> 
> In Dockerfile recipe (image creation), by default DATADIR is in `/var/lib/postgresql/12/main` . So, the directory has content (in the image). When user uses named-volume and bind into this default DATADIR: if the named-volume is empty, the existing content is copied into the named volume (by docker automatically). So the default collation/settings is used and can't be changed. The entrypoint script will skip initialization phase because DATADIR already had content.
> # Datadir maintained by user
> 
> If user reuse existing named-volume that already has DATADIR content, then we should not delete it. The container entry point script can only delete it if RECREATE_DIR is true (the default will be false). DATADIR maintained by user can apply in many scenarios such as:
> 
>     * User uses our image initially with empty named-volume. At the end, his named-volume has content. If he change image tag but keep the major version (e.g postgis:12.0 into postgis:12.1), he should be able to use the same datadir in his named-volume.
> 
>     * User migrated his own datadir from somewhere and mount the volume as the container's DATADIR. If it is the same major version, it should work and we don't delete it.
> 
>     * User migrated his own datadir, or uses initialized named-volume, but from a different postgres/postgis major version. In this case, we should not delete datadir. But rather, hang gracefully with nice log that says why we can't use the datadir (the error message from postgres). It needs to hang so user can get into the container and resolves it. If we don't make it hang, the container will be in an endless loop of restart.
> 
> 
> # User gave non default DATADIR locations
> 
> If the user set DATADIR in a non default locations (other than `/var/lib/postgresql/12/main`), with empty named-volume, then the DATADIR is empty. So entrypoint must initialize DATADIR according to the chosen conf.
> 
> So, to comment on your statement here:
> 
> > I am very much against leaving the container hanging. It should handle both scenarios as before.
> 
> We're not going to use the current commits in this PR, it will be updated according to the spec that we will agree. What I propose is for the image to handle as above (and then you can add if I miss something). If the scenario is the DATADIR is not empty and postgres can't start, then the container should hang so user can read the logs and resolves the situation. Because we don't know exactly how to resolve it (should we delete the DATADIR or not). User had to give explicit command or settings like RECREATE_DATADIR.
> 
> > Let's just delete environment variable EXISTING_DATA_DIR, rather change it into FORCE_REINITIALIZATION or something (default to False). This is to clearly express the intent, because: If the datadir is not from the same major version of postgres, entrypoint should fail and leave it hanging so the container is up, but unhealthy. That way, user can get inside the container and fix the conflicts themselves (use another dir or change image version).
> 
> This statement above should have been a single point of idea, my bad, it was separated as point 3 and 4.
> 
> To compare it in a table, the difference will be:
> No 	Before 	After
> 1 	Entry point script delete DATADIR if lock file does not exists 	Entry point script will delete DATADIR only if RECREATE_DATADIR is set to true
> 2 	Entry point script will create lock file when DATADIR is initialized 	No lockfile
> 3 	Entry point script will still reinitialize DATADIR at first run, because lock file doesn't exists even if using DATADIR prebuilt by image 	Entry point script will only initialize if DATADIR is empty or RECREATE_DATADIR is true
> 4 	EXISTING_DATADIR is used as flag to tell that DATADIR exists 	Not using EXISTING_DATADIR because it had no purpose in this implementation
> 5 	EXISTING_DATADIR default to False, so user who upgrade the image pre EXISTING_DATADIR options will not have lock file, and it's data were deleted 	Never delete data unless explicit consent as settings were given or user delete it manually

@lucernae I agree with the logic. You can go ahead and implement and I will commenton the PR

lucernae committed:
Fix issue #226

- Refactor entrypoint process to only delete DATADIR when
  RECREATE_DATADIR is explicitly set to TRUE
- Refactor config to use DATADIR from environment variable
- Update README

NyakudyaA commented on setup-database.sh:
```diff
@@ -79,7 +56,7 @@ for db in $(echo ${POSTGRES_DBNAME} | tr ',' ' '); do
         if [[  ${RESULT} -eq 0 ]]; then
             echo "Create db ${db}"
-            su - postgres -c "createdb -O ${POSTGRES_USER}  ${db}"
+            su - postgres -c "createdb -O ${POSTGRES_USER} -E ${DEFAULT_ENCODING} --lc-collate=${DEFAULT_COLLATION} --lc-ctype=${DEFAULT_CTYPE} -T template0 ${db}"
```
If we initialize the DataDIR with locales we should not require the creation to use locales as well. This will be duplication ?

lucernae replied:
I checked it with newly created datadir, previously the collation of the initdb parameter doesn’t change the collation of the database created. But it works if I explicitly set it like this.
I will add new unittests to demonstrate this.

NyakudyaA commented on setup-database.sh:
```diff
@@ -2,43 +2,20 @@
+    rm -rf ${DATADIR}/*
+    chown -R postgres:postgres ${DATADIR}
+    echo "Initializing with command:"
+    command="$INITDB -U postgres -E ${DEFAULT_ENCODING} --lc-collate=${DEFAULT_COLLATION} --lc-ctype=${DEFAULT_CTYPE} --wal-segsize=${WAL_SEGSIZE} -D ${DATADIR}"
```
I thought we could also add an extra env variable for users to add to initialize the data directory ie POSTGRES_INITDB_EXTRA
the default could be empty but users could expand it example https://www.postgresql.org/docs/12/app-initdb.html
ie POSTGRES_INITDB_EXTRA=--waldir=/data
```
command="$INITDB -U postgres -E ${DEFAULT_ENCODING} --lc-collate=${DEFAULT_COLLATION} --lc-ctype=${DEFAULT_CTYPE} --wal-segsize=${WAL_SEGSIZE} -D ${DATADIR} ${POSTGRES_INITDB_EXTRA}"
```
where example POSTGRES_INITDB_EXTRA=--waldir=/data

lucernae replied:
Good idea! Might as well do that too.

NyakudyaA replied:
Did you add this option. I think it would also help with the wal being full because you could set it to a different folder

lucernae committed:
Refactor tests to use python

lucernae committed:
Add collations tests

lucernae commented:
@NyakudyaA I added scenario about collations here:
[scenario_tests/collations/docker-compose.yml](https://github.com/kartoza/docker-postgis/pull/228/files#diff-2a36091b7eb7116ef74cac84ec1178c2)
 
Following up our discussion [here](https://github.com/kartoza/docker-postgis/pull/228#pullrequestreview-385594887):

So there will be at least two use case and the line addition in that file is for one of them:

- If user do not won't to recreate the whole datadir (RECREATE_DATADIR is false), but only creates new database with given collation (this is why I add that line and unittest to check that)
- `DEFAULT_COLLATION` and `DEFAULT_CTYPE` should be applied when recreating datadir (the codebase already covered that).

If we don't need the first use case, then we can just delete the line, I guess? Let me know.

lucernae committed:
Change travis config to omit docker image sharing

NyakudyaA commented:
> @NyakudyaA I added scenario about collations here:
> [scenario_tests/collations/docker-compose.yml](https://github.com/kartoza/docker-postgis/pull/228/files#diff-2a36091b7eb7116ef74cac84ec1178c2)
> 
> Following up our discussion [here](https://github.com/kartoza/docker-postgis/pull/228#pullrequestreview-385594887):
> 
> So there will be at least two use case and the line addition in that file is for one of them:
> 
>     * If user do not won't to recreate the whole datadir (RECREATE_DATADIR is false), but only creates new database with given collation (this is why I add that line and unittest to check that)
> 
>     * `DEFAULT_COLLATION` and `DEFAULT_CTYPE` should be applied when recreating datadir (the codebase already covered that).
> 
> 
> If we don't need the first use case, then we can just delete the line, I guess? Let me know.

* If a user creates a data directory with collations then the databases created should not specify the collation because it will be inherited from the data-dir.
* If a user chooses to use the default data dir with default collation and he specifies a different collation for the database, are we ever going to have this situation?. This is why I believe that line should not be there because we either have to choose data dir collation which is better

lucernae commented:
> If a user creates a data directory with collations then the databases created should not specify the collation because it will be inherited from the data-dir.

Yes, this is already covered in the codebase. (not in this PR)

>  If a user chooses to use the default data dir with default collation and he specifies a different collation for the database, are we ever going to have this situation?. This is why I believe that line should not be there because we either have to choose data dir collation which is better

What I’m trying to say is, without that line, even if you specify DEFAULT_COLLATION, the database created with current datadir won’t follow this. If this is acceptable than we don’t have to put that line. So if the line was omitted, to use DEFAULT_COLLATION settings, the user needs to (one of the use case):
- Specify RECREATE_DATADIR true to destroy current datadir, and recreate new with new collation settings as specified by DEFAULT_COLLATION

or

- With empty named-volume, use PGDATA to specify non default locations, so datadir is empty and postgres entry point script will have to create new datadir with specified collation settings.


The DEFAULT_COLLATION settings is not going to work if the line was omitted in this case:

- Empty named-volume, using default DATADIR/PGDATA location. The newly created database by entrypoint script (by default named `gis`), will have default collation (C.UTF-8) and DEFAULT_COLLATION settings is ignored. Because the datadir were copied into the named-volume by docker engine at the time of container creations, so when entrypoint script runs, named-volume already had contents.

I’m just unsure you want it to behave like the last case here. But if this is intended, we omit that line and improve the README to tell user that DEFAULT_COLLATION only have effects when PGDATA location is empty or RECREATE_DATADIR is true.

NyakudyaA commented:
> > If a user creates a data directory with collations then the databases created should not specify the collation because it will be inherited from the data-dir.
> 
> Yes, this is already covered in the codebase. (not in this PR)
> 
> > If a user chooses to use the default data dir with default collation and he specifies a different collation for the database, are we ever going to have this situation?. This is why I believe that line should not be there because we either have to choose data dir collation which is better
> 
> What I’m trying to say is, without that line, even if you specify DEFAULT_COLLATION, the database created with current datadir won’t follow this. If this is acceptable than we don’t have to put that line. So if the line was omitted, to use DEFAULT_COLLATION settings, the user needs to (one of the use case):
> 
> * Specify RECREATE_DATADIR true to destroy current datadir, and recreate new with new collation settings as specified by DEFAULT_COLLATION
> 
> or
> 
> * With empty named-volume, use PGDATA to specify non default locations, so datadir is empty and postgres entry point script will have to create new datadir with specified collation settings.
> 
> The DEFAULT_COLLATION settings is not going to work if the line was omitted in this case:
> 
> * Empty named-volume, using default DATADIR/PGDATA location. The newly created database by entrypoint script (by default named `gis`), will have default collation (C.UTF-8) and DEFAULT_COLLATION settings is ignored. Because the datadir were copied into the named-volume by docker engine at the time of container creations, so when entrypoint script runs, named-volume already had contents.
> 
> I’m just unsure you want it to behave like the last case here. But if this is intended, we omit that line and improve the README to tell user that DEFAULT_COLLATION only have effects when PGDATA location is empty or RECREATE_DATADIR is true.

Let me test the PR and will then revert. But looking good

lucernae committed:
Change tests for collation default

- Following up PR #228 about environment variable behaviour

lucernae committed:
Add tests for datadir initializations

lucernae committed:
Update Travis config

lucernae committed:
Improve replication testing

lucernae commented:
@NyakudyaA I've implemented what we discussed in this last commit:
- Improve docs to point out initialization parameter
- Only create new cluster when `RECREATE_DATADIR` is true OR the `DATADIR` provided is empty. Otherwise try to use existing `DATADIR`
- Omit collation settings from database creation
- Make 3 separate service unittests (collation, datadir_init, replication)

NyakudyaA commented:
LGTM @lucernae

lucernae commented:
Thanks @NyakudyaA I will merge and backport to 11
